### PR TITLE
Add range search capability to indexes with AnnIterator support

### PR DESCRIPTION
**Pull request by zhengbuqian**
tested using sparse index.

issue: https://github.com/zilliztech/knowhere/issues/193
/kind: feature

### CODE REVIEW

**Comment by zhengbuqian** [comment]
/hold
It won't compile with swig for now

**Comment by zhengbuqian** [comment]
> /hold It won't compile with swig for now

fixing in https://github.com/zilliztech/knowhere/pull/509, will unhold after that is merged.

File: src/common/comp/brute_force.cc [hunk 1]
```diff
@@ -341,12 +348,27 @@ BruteForce::RangeSearch(const DataSetPtr base_dataset, const DataSetPtr query_da
...
+
     std::vector<folly::Future<Status>> futs;
     futs.reserve(nq);
     for (int i = 0; i < nq; ++i) {
         futs.emplace_back(pool->push([&, index = i] {
+            if (is_sparse) {
+                auto cur_query = (const sparse::SparseRow<float>*)xq + index;
+                auto xb_sparse = (const sparse::SparseRow<float>*)xb;
+                for (int j = 0; j < nb; ++j) {
+                    if (!bitset.empty() && bitset.test(j)) {
```
[the remark this answers is not in the document]
Comment (zhengbuqian) [reply src/common/comp/brute_force.cc:360]: I find no `valid` method? can you clarify?

File: src/common/comp/brute_force.cc [hunk 2]
```diff
@@ -341,12 +348,27 @@ BruteForce::RangeSearch(const DataSetPtr base_dataset, const DataSetPtr query_da
...
+                        continue;
+                    }
+                    auto dist = cur_query->dot(xb_sparse[j]);
+                    if (dist > radius && dist <= range_filter) {
```
[the remark this answers is not in the document]
Comment (zhengbuqian) [reply src/common/comp/brute_force.cc:364]: yes we support only IP metric for sparse vectors. L2 can technically be computed but is much less useful for sparse vectors, mainly due to the curse of dimensionality.

File: include/knowhere/index_node.h [hunk 3]
```diff
@@ -70,11 +76,71 @@ class IndexNode : public Object {
...
+                auto [id, dist] = it->Next();
+                // too close
+                if (similarity_metric ? dist >= range_filter : dist <= range_filter) {
+                    continue;
+                }
+                // too far
+                if (similarity_metric ? dist <= radius : dist >= radius) {
+                    break;
+                }
+                result_id_array[idx].push_back(id);
```
[the remark this answers is not in the document]
Comment (zhengbuqian) [reply include/knowhere/index_node.h:120]: the comments are correct. I wrote the code based on notes in https://milvus.io/docs/v2.3.x/within_range.md#Distance-metrics-influence, which is wrong and will be updated.

File: src/common/comp/brute_force.cc (hunk 1 above)
[the remark this answers is not in the document]
Comment (alexanderguzhva) [reply src/common/comp/brute_force.cc:360]: My bad, I mistook this bitset implementation with another implementation that has `valid()` method

**Pull request merged**